### PR TITLE
Wikidata

### PR DESCRIPTION
**Pull request by nayamamura**
Wikidata Q&A over CSQA

### CODE REVIEW

File: tool/autoqa/wikidata/csqa-converter.js [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+    async _filterDomainQAPairs(canonical) {
+        const basePath = path.join(this._input_dir, this._dataset)
+        let cnt = 0;
+        const qaPairs = [];
+        for (const dir of (await this. _readSync(fs.readdir, basePath))) {
+            for (const file of (await this. _readSync(fs.readdir, path.join(basePath, dir)))) {
+                const dialog = JSON.parse((await this. _readSync(fs.readFile, path.join(basePath, dir, file))));
+                let userTurn;
+                for (const turn in dialog) {
+                    const speaker = dialog[turn].speaker;
```
Comment (rayslxu) [inline tool/autoqa/wikidata/csqa-converter.js:78]: there are quite some nested loop/conditions, probably consider create a separate function such as `filterTurnByDomain` to make this function easier to read

File: tool/autoqa/wikidata/csqa-converter.js [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+
+        const filterTable = new Ast.FilterExpression(null, invocationTable, filter, null);
+        let param;
+        if (activeSet[1] === activeSet[4]) {
+            param = [(await this.getArgName(activeSet[2], activeSet[1]))];
+        } else {
+            param = [(await this.getArgName(activeSet[2], activeSet[1])), (await this.getArgName(activeSet[5], activeSet[4]))];
+        }
+        // Return projected table.
+        return (new Ast.ProjectionExpression(null, filterTable, param, [], []));
```
Comment (rayslxu) [inline tool/autoqa/wikidata/csqa-converter.js:213]: Missing one param (type check warning by IDE)

File: tool/preprocess-entity-dataset.js [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+class EntityPreprocessor {
+    constructor(options) {
+        this._locale = options.locale;
+        this._tokenizer = I18N.get(options.locale).getTokenizer();
+    }
+
+    async process(fins, fout) {
+        const values = [];
+        for (const fin of fins) {
+            let input = JSON.parse(await util.promisify(fs.readFile)(fin), {encoding: 'utf8'});
```
Comment (rayslxu) [inline tool/preprocess-entity-dataset.js:36]: all this large file preprocessing is probably better to be done with streams... but this has lower priority to fix

File: tool/autoqa/wikidata/csqa-converter.js [hunk 4]
```diff
@@ -0,0 +1,398 @@
...
+        return (new Ast.InvocationExpression(null, new Ast.Invocation(null, selector, canonical, [], null), null));
+    }
+
+    async getSingleFilter(activeSet, negate) {
+        const param = await this.getArgName(activeSet[0], activeSet[1]);
+        const arg = await this.getArgValue(activeSet[0])
+        if (!arg) return;
+
+        const ttValue = new Ast.Value.String(arg);
+        const exp = new Ast.BooleanExpression.Atom(null, param, '=~', ttValue);
```
Comment (rayslxu) [inline tool/autoqa/wikidata/csqa-converter.js:145]: all filters are string filters?

**Review by rayslxu — CHANGES_REQUESTED** [review]
I've updated the serialization of tokenized thingtalk - you need to tokenize the utterance, not the code to extract entities. 
There are quite some of lint and coding style issues. I recommend you to setup type checking and lint in your IDE to move forward. 
There are also a few technical questions I raised.

File: package-lock.json [hunk 5]
```diff
@@ -2213,15 +2212,21 @@
       "dev": true
     },
     "fs-extra": {
-      "version": "9.1.0",
-      "resolved": "https://registry.npmjs.org/fs-extra/-/fs-extra-9.1.0.tgz",
-      "integrity": "sha512-hcg3ZmepS30/7BSFqRvoo3DOMQu7IjqxO5nCDt+zM9XWjb33Wg7ziNT+Qvqbuc3+gWpzO02JubVyk2G4Zvo1OQ==",
-      "dev": true,
+      "version": "9.0.1",
```
[the remark this answers is not in the document]
Comment (nayamamura) [reply package-lock.json]: Reverted.

File: tool/autoqa/wikidata/csqa-converter.js [hunk 6]
```diff
@@ -0,0 +1,398 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Author: Naoki Yamamura <yamamura@cs.stanford.edu>
+"use strict";
+
+
+const fs = require('fs');
+const os = require('os');
+const _ = require('lodash');
```
[the remark this answers is not in the document]
Comment (nayamamura) [reply tool/autoqa/wikidata/csqa-converter.js]: Removed.

File: tool/autoqa/wikidata/preprocess-data.js [hunk 7]
```diff
@@ -0,0 +1,385 @@
...
+        // Required CSQA Wikidata json files
+        assert(fs.existsSync(path.join(this._input_dir, 'filtered_property_wikidata4.json')));
+        assert(fs.existsSync(path.join(this._input_dir, `items_wikidata_n.json`)));
+        assert(fs.existsSync(path.join(this._input_dir, `wikidata_short_1.json`)));
+        assert(fs.existsSync(path.join(this._input_dir, `wikidata_short_2.json`)));
+        assert(fs.existsSync(path.join(this._input_dir, `comp_wikidata_rev.json`)));
+
+        await loadSchemaOrgManifest(this._schemaorgManifest, this._schemaorgProperties);
+        for (const idx in this._domains) {
+            this._propery_pathes = [
```
[the remark this answers is not in the document]
Comment (nayamamura) [reply tool/autoqa/wikidata/preprocess-data.js]: updated.

File: tool/autoqa/wikidata/csqa-converter.js (hunk 4 above)
Comment (nayamamura) [reply tool/autoqa/wikidata/csqa-converter.js]: updated but want to discuss a bit.

File: tool/autoqa/wikidata/manual-annotations.js [hunk 8]
```diff
@@ -44,10 +44,111 @@ const PROPERTY_TYPE_OVERRIDE = {
...
+    'P113': Type.Location, // airline hub
+    'P81': new Type.Entity('org.wikidata:railway_line'), // connecting line
+    'P197': Type.Location, // adjacent station
+    'P403': new Type.Entity('org.wikidata:mouth_of_the_watercourse'), // mouth of the watercourse
+    'P734': new Type.Entity('org.wikidata:name'), // family name
+    'P2632': Type.Location, // place of detention
+    'P1686': new Type.Entity('org.wikidata:work'), // for work
+    'P768': new Type.Entity('org.wikidata:district'), // electoral district
+    'P413': new Type.Entity('org.wikidata:position'), // position played on team / speciality
+    'P2554': new Type.Entity('org.wikidata:designer'), // production designer
```
[the remark this answers is not in the document]
Comment (nayamamura) [reply tool/autoqa/wikidata/manual-annotations.js]: this was more of experiment to see if mapping type would help or not. reverted.

File: tool/preprocess-entity-dataset.js (hunk 3 above)
Comment (nayamamura) [reply tool/preprocess-entity-dataset.js]: After clean up this file is not required. Removed.

**Review by rayslxu — COMMENTED** [review]
All looks reasonable to me now. I will wait for the starter code to test run it.

**Comment by rayslxu** [comment]
There is still a lot of work needed for this PR still. Will reopen once ready.

**Pull request closed without merging**